staging/src/k8s.io/api/apps/v1/types.go

```diff
@@ -867,29 +873,35 @@ type ReplicaSetSpec struct {
 
 	// Template is the object that describes the pod that will be created if
 	// insufficient replicas are detected.
-	// More info: https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller#pod-template
+	// More info: https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-template
 	// +optional
 	Template v1.PodTemplateSpec `json:"template,omitempty" protobuf:"bytes,3,opt,name=template"`
 }
 
 // ReplicaSetStatus represents the current status of a ReplicaSet.
 type ReplicaSetStatus struct {
-	// Replicas is the most recently observed number of replicas.
-	// More info: https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller/#what-is-a-replicationcontroller
+	// Replicas is the most recently observed number of non-terminating pods.
+	// More info: https://kubernetes.io/docs/concepts/workloads/controllers/replicaset
 	Replicas int32 `json:"replicas" protobuf:"varint,1,opt,name=replicas"`
 
-	// The number of pods that have labels matching the labels of the pod template of the replicaset.
+	// The number of non-terminating pods that have labels matching the labels of the pod template of the replicaset.
 	// +optional
 	FullyLabeledReplicas int32 `json:"fullyLabeledReplicas,omitempty" protobuf:"varint,2,opt,name=fullyLabeledReplicas"`
 
-	// readyReplicas is the number of pods targeted by this ReplicaSet with a Ready Condition.
+	// The number of non-terminating pods targeted by this ReplicaSet with a Ready Condition.
 	// +optional
 	ReadyReplicas int32 `json:"readyReplicas,omitempty" protobuf:"varint,4,opt,name=readyReplicas"`
 
-	// The number of available replicas (ready for at least minReadySeconds) for this replica set.
+	// The number of available non-terminating pods (ready for at least minReadySeconds) for this replica set.
 	// +optional
 	AvailableReplicas int32 `json:"availableReplicas,omitempty" protobuf:"varint,5,opt,name=availableReplicas"`
 
+	// The number of terminating pods (have non-null .metadata.deletionTimestamp) for this replica set.
+	//
+	// This is an alpha field. Enable DeploymentPodReplacementPolicy to be able to use this field.
+	// +optional
+	TerminatingReplicas int32 `json:"terminatingReplicas,omitempty" protobuf:"varint,7,opt,name=terminatingReplicas"`
```

Comment:
this must be *int32, otherwise we immediately start persisting changes into etcd even when the feature is disabled (see round-trip errors on 1.31 and 1.30 fixtures), and we cannot tell the difference between a server / controller that is not populating this field and one that is asserting that there are no terminating replicas

Reply:
this ripples out to all the handling code which should set to non-nil if the feature is enabled, set to nil if the feature is not enabled, and only populate deployment status if all the replicasets being summed have non-nil status values for this field

Reply:
We had this conversation in the KEP review: 
- https://github.com/kubernetes/enhancements/pull/4357#discussion_r1476726816
- https://github.com/kubernetes/enhancements/pull/4357#discussion_r1475537038

So, I was under the impression that is okay, since we did that with the older fields (e.g. https://github.com/kubernetes/kubernetes/pull/100842): https://github.com/kubernetes/kubernetes/blob/af41aa1d9f7f347c85aba629963941a8ce6a95fd/staging/src/k8s.io/api/apps/v1/types.go#L313-L315

Reply:
I see, in that case I will change the fields. If we are okay that the status fields will be inconsistent (similar non-ptr and ptr fields)